README.md
# vscode-localize README

This is the README for your extension "vscode-localize". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "vscode-localize" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/_extension.ts
import * as vscode from 'vscode';

// this method is called when vs code is activated
export function activate(context: vscode.ExtensionContext) {

	console.log('decorator sample is activated');

	let timeout: NodeJS.Timer | undefined = undefined;

	// create a decorator type that we use to decorate small numbers
	const inactiveTranslationDecorationType = vscode.window.createTextEditorDecorationType({
		after: {
			contentText: 'Stäng fönster',
			color: '#6699cc'
		},
		opacity: '0',
		letterSpacing: '-1em',
		overviewRulerColor: 'blue',
		overviewRulerLane: vscode.OverviewRulerLane.Right
	});

	const activeTranslationDecorationType = vscode.window.createTextEditorDecorationType({
		after: {
			contentText: ':Stäng fönster',
			color: '#6699cc'
		},
		opacity: '0.35',
		overviewRulerColor: 'blue',
		overviewRulerLane: vscode.OverviewRulerLane.Right
	});

	let activeEditor = vscode.window.activeTextEditor;

	function updateDecorations() {
		if (!activeEditor) {
			return;
		}
		const regEx = /\$t\('([^']+)'\)/g;
		const text = activeEditor.document.getText();
		const inactiveTranslations: vscode.DecorationOptions[] = [];
		const activeTranslations: vscode.DecorationOptions[] = [];

		let offset = activeEditor.document.offsetAt(activeEditor.selection.active);
		
		console.log('Offset:', offset)

		let match;
		while (match = regEx.exec(text)) {
			let index = match[0].indexOf(match[1]) + match.index;

			const startPos = activeEditor.document.positionAt(index);
			const endPos = activeEditor.document.positionAt(index + match[1].length);
			const decoration = {
				range: new vscode.Range(startPos, endPos),
				hoverMessage: 'Number **' + match[0] + '**'
			};
			const cursorInside = offset >= index && offset <= (index + match[1].length)

			if (cursorInside) {
				activeTranslations.push(decoration);
			} else {
				inactiveTranslations.push(decoration);
			}
		}
		activeEditor.setDecorations(inactiveTranslationDecorationType, inactiveTranslations);
		activeEditor.setDecorations(activeTranslationDecorationType, activeTranslations);
	}

	function triggerUpdateDecorations() {
		if (timeout) {
			clearTimeout(timeout);
			timeout = undefined;
		}
		timeout = setTimeout(updateDecorations, 50);
	}

	if (activeEditor) {
		triggerUpdateDecorations();
	}

	vscode.window.onDidChangeActiveTextEditor(editor => {
		activeEditor = editor;
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	vscode.window.onDidChangeTextEditorSelection(editor => {
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	vscode.workspace.onDidChangeTextDocument(event => {
		if (activeEditor && event.document === activeEditor.document) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

}

src/extension.1.ts
import * as vscode from 'vscode';

interface TranslationMap {
	[key:string]: string;
}
interface DecorationMap {
	[key:string]: {
		active: vscode.TextEditorDecorationType,
		inactive: vscode.TextEditorDecorationType
	};
}
interface TranslationDecorator {
	[key:string]: any;
}
interface TranslationDecoratorMap {
	[key:string]: any[];
}

// this method is called when vs code is activated
export function activate(context: vscode.ExtensionContext) {

	const translations: TranslationMap = {
		'g8ck3s': 'Hello, world!',
		'v90j12': 'Close',
		'k0j2fg': 'dispatch'
	};

	console.log('decorator sample is activated');

	let timeout: NodeJS.Timer | undefined = undefined;

	let decorators: DecorationMap = {};

	for (let key in translations) {
		const active = vscode.window.createTextEditorDecorationType({
			after: {
				contentText: translations[key],
				color: '#6699cc'
			},
			opacity: '0.35',
			overviewRulerColor: 'blue',
			overviewRulerLane: vscode.OverviewRulerLane.Right
		});
		const inactive = vscode.window.createTextEditorDecorationType({
			after: {
				contentText: translations[key],
				color: '#6699cc'
			},
			opacity: '0',
			letterSpacing: '-1em',
			overviewRulerColor: 'blue',
			overviewRulerLane: vscode.OverviewRulerLane.Right
		});

		decorators[key] = {
			active,
			inactive
		};
	}

	let activeEditor = vscode.window.activeTextEditor;

	function updateDecorations() {
		if (!activeEditor) {
			return;
		}
		const regEx = /\$t\('([^']+)'\)/g;
		const text = activeEditor.document.getText();

		const translationDecorations: TranslationDecoratorMap = {};

		let offset = activeEditor.document.offsetAt(activeEditor.selection.active);

		let match;
		while (match = regEx.exec(text)) {
			if (!translations[match[1]]) {
				continue;
			}

			let index = match[0].indexOf(match[1]) + match.index;

			const startPos = activeEditor.document.positionAt(index);
			const endPos = activeEditor.document.positionAt(index + match[1].length);
			const cursorInside = offset >= index && offset <= (index + match[1].length)
			
			const decoration = {
				range: new vscode.Range(startPos, endPos),
				hoverMessage: 'Number **' + match[0] + '**',
				active: cursorInside
			};

			if (!translationDecorations[match[1]]) {
				translationDecorations[match[1]] = [];
			}

			translationDecorations[match[1]].push(decoration);
		}

		let decorationKeys = Object.keys(decorators);

		let disposeKeys = decorationKeys.filter(key => !translationDecorations[key]);

		for (let key in translationDecorations) {
			const decorations = translationDecorations[key];
			activeEditor.setDecorations(decorators[key].active, decorations);
		}

		for (let key of disposeKeys) {
			decorators[key].active.dispose();
		}
	}

	function triggerUpdateDecorations() {
		if (timeout) {
			clearTimeout(timeout);
			timeout = undefined;
		}
		timeout = setTimeout(updateDecorations, 50);
	}

	if (activeEditor) {
		triggerUpdateDecorations();
	}

	vscode.window.onDidChangeActiveTextEditor(editor => {
		activeEditor = editor;
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	vscode.window.onDidChangeTextEditorSelection(editor => {
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	vscode.workspace.onDidChangeTextDocument(event => {
		if (activeEditor && event.document === activeEditor.document) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

}

src/localization.ts
import { workspace, window } from "vscode";
import { readdir } from "fs";
import { join as joinPath } from "path";

export function loadLocalizations (callback?: Function) {
  console.log('Loading localizations...');
  if (!workspace.workspaceFolders || workspace.workspaceFolders.length <= 0) {
    return;
  }

  const configLocalesFolder: string | undefined = config.get('localesDirectory');

  if (configLocalesFolder === undefined) {
    window.showWarningMessage('No locales folder found.');
    return;
  }

  let primaryFolder = workspace.workspaceFolders[0];
  let localesFolder = joinPath(primaryFolder.uri.fsPath, configLocalesFolder);

  readdir(localesFolder, (err, files) => {
    if (err) {
      throw err;
    }

    loadedLocales = files.map(file => file.replace(/.json$/, ''));

    let localePath = path.join(localesFolder, `${currentLocale}.json`);

    for (let key in decorators) {
      decorators[key].active.dispose();
      decorators[key].inactive.dispose();
    }

    decorators = {};

    vscode.workspace.openTextDocument(localePath).then((document) => {
      let text = document.getText();
      let localizations = JSON.parse(text);

      translations = localizations;
      codeLensProvider.setTranslations(translations);

      for (let key in translations) {
        const active = vscode.window.createTextEditorDecorationType({
          after: {
            contentText: ` ${translations[key]}`,
            color: '#6699cc'
          },
          opacity: '0.35',
          overviewRulerColor: 'blue',
          overviewRulerLane: vscode.OverviewRulerLane.Right
        });
        const inactive = vscode.window.createTextEditorDecorationType({
          after: {
            contentText: translations[key],
            color: '#6699cc'
          },
          opacity: '0',
          letterSpacing: '-1em',
          overviewRulerColor: 'blue',
          overviewRulerLane: vscode.OverviewRulerLane.Right
        });
    
        decorators[key] = {
          active,
          inactive
        };
      }

      if (callback) {
        callback();
      }
    });
  });
}

src/LocaleCodeLensProvider.ts
import {
  CodeLensProvider,
  TextDocument,
  CodeLens,
  Range,
  Command,
  window
} from "vscode";

import { TranslationMap } from "./interfaces";

class LocaleCodeLensProvider implements CodeLensProvider {
  private translations: TranslationMap = {};

  setTranslations (translations: TranslationMap) {
    this.translations = translations;
  }

  // Each provider requires a provideCodeLenses function which will give the various documents
  // the code lenses
  async provideCodeLenses(document: TextDocument): Promise<CodeLens[]> {
    let activeEditor = window.activeTextEditor;
    
    if (!activeEditor) {
      return [];
    }

    const regEx = /\$t\('([^']+)'\)/g;
		const text = activeEditor.document.getText();

    let lenses: CodeLens[] = [];
    
		let match;
		while (match = regEx.exec(text)) {
			if (!this.translations[match[1]]) {
				continue;
			}

			let index = match[0].indexOf(match[1]) + match.index;

			const startPos = activeEditor.document.positionAt(index);
			const endPos = activeEditor.document.positionAt(index + match[1].length);

      console.log('Match:', match[1]);
      
      const command: Command = {
        command: "extension.openTranslation",
        arguments: [match[1]],
        title: "Edit translation"
      };
      
			lenses.push(new CodeLens(new Range(startPos, endPos), command));
		}

    return lenses;
  }
}

export default LocaleCodeLensProvider;

src/extension.ts
import { ExtensionContext, window, WorkspaceConfiguration, languages, workspace, OverviewRulerLane, Range, commands } from 'vscode';

import { join as joinPath } from 'path';
import { readdir } from 'fs';

import registerCommands from './registerCommands';

import LocaleCodeLensProvider from './LocaleCodeLensProvider';

import { switchLocale } from './commands';

import {
	TranslationMap,
	DecorationMap,
	TranslationDecoratorMap
} from './interfaces';

export function activate(context: ExtensionContext) {
	console.log('VS Code Localize activated');

	registerCommands(context);
	
	let loadedLocales: string[] = [];
	let currentLocale: string;
	
	let translations: TranslationMap = {};
	let decorators: DecorationMap = {};
	

	let activeEditor = window.activeTextEditor;
	let config: WorkspaceConfiguration;
	
	let timeout: NodeJS.Timer | undefined = undefined;
	
	let docSelector = {
		language: 'javascript',
		scheme: 'file'
	};
	
	let codeLensProvider = new LocaleCodeLensProvider();
	let codeLensProviderDisposable = languages.registerCodeLensProvider(
    docSelector,
    codeLensProvider
  );

  // Push the command and CodeLens provider to the context so it can be disposed of later
  context.subscriptions.push(codeLensProviderDisposable);

	function loadConfiguration () {
		config = workspace.getConfiguration('localize');
		const configDefaultLocale: string | undefined = config.get('defaultLocale');

		if (!configDefaultLocale) {
			return;
		}
		currentLocale = configDefaultLocale;
	}

	function loadLocalizations (callback?: Function) {
		console.log('Loading localizations...');
		if (!workspace.workspaceFolders || workspace.workspaceFolders.length <= 0) {
			return;
		}

		const configLocalesFolder: string | undefined = config.get('localesDirectory');

		if (configLocalesFolder === undefined) {
			window.showWarningMessage('No locales folder found.');
			return;
		}

		let primaryFolder = workspace.workspaceFolders[0];
		let localesFolder = joinPath(primaryFolder.uri.fsPath, configLocalesFolder);

		readdir(localesFolder, (err, files) => {
			if (err) {
				throw err;
			}

			loadedLocales = files.map(file => file.replace(/.json$/, ''));

			let localePath = joinPath(localesFolder, `${currentLocale}.json`);

			for (let key in decorators) {
				decorators[key].active.dispose();
				decorators[key].inactive.dispose();
			}

			decorators = {};

			workspace.openTextDocument(localePath).then((document) => {
				let text = document.getText();
				let localizations = JSON.parse(text);

				translations = localizations;
				codeLensProvider.setTranslations(translations);

				for (let key in translations) {
					const active = window.createTextEditorDecorationType({
						after: {
							contentText: ` ${translations[key]}`,
							color: '#6699cc'
						},
						opacity: '0.35',
						overviewRulerColor: 'blue',
						overviewRulerLane: OverviewRulerLane.Right
					});
					const inactive = window.createTextEditorDecorationType({
						after: {
							contentText: translations[key],
							color: '#6699cc'
						},
						opacity: '0',
						letterSpacing: '-1em',
						overviewRulerColor: 'blue',
						overviewRulerLane: OverviewRulerLane.Right
					});
			
					decorators[key] = {
						active,
						inactive
					};
				}

				if (callback) {
					callback();
				}
			});
		});
	}

	function updateDecorations(allEditors?: boolean) {
		let editors = [activeEditor];

		if (allEditors) {
			editors = window.visibleTextEditors;
		}

		for (let editor of editors) {
			if (!editor) {
				return;
			}
	
			for (let key in decorators) {
				editor.setDecorations(decorators[key].active, []);
				editor.setDecorations(decorators[key].inactive, []);
			}
	
			const regEx = /\$t\('([^']+)'\)/g;
			const text = editor.document.getText();
	
			const translationDecorations: TranslationDecoratorMap = {};
	
			let offset = editor.document.offsetAt(editor.selection.active);
	
			let match;
			while (match = regEx.exec(text)) {
				if (!translations[match[1]]) {
					continue;
				}
	
				let index = match[0].indexOf(match[1]) + match.index;
	
				const startPos = editor.document.positionAt(index);
				const endPos = editor.document.positionAt(index + match[1].length);
				const cursorInside = offset >= index && offset <= (index + match[1].length);
				
				const decoration = {
					range: new Range(startPos, endPos),
					hoverMessage: `Localization key **${match[1]}**`,
					active: cursorInside
				};
	
				if (!translationDecorations[match[1]]) {
					translationDecorations[match[1]] = [];
				}
	
				translationDecorations[match[1]].push(decoration);
			}
	
			let decorationKeys = Object.keys(decorators);
	
			let disposeKeys = decorationKeys.filter(key => !translationDecorations[key]);
	
			for (let key in translationDecorations) {
				const decorations = translationDecorations[key];
				const active = decorations.filter(decoration => decoration.active);
				const inactive = decorations.filter(decoration => !decoration.active);
				editor.setDecorations(decorators[key].active, active);
				editor.setDecorations(decorators[key].inactive, inactive);
			}
	
			for (let key of disposeKeys) {
				editor.setDecorations(decorators[key].active, []);
				editor.setDecorations(decorators[key].inactive, []);
			}
		}
	}

	function triggerUpdateDecorations() {
		if (timeout) {
			clearTimeout(timeout);
			timeout = undefined;
		}
		
		timeout = setTimeout(updateDecorations, 50);
	}

	if (activeEditor) {
		triggerUpdateDecorations();
	}
	
	loadConfiguration();

	window.onDidChangeActiveTextEditor(editor => {
		activeEditor = editor;
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	window.onDidChangeTextEditorSelection(editor => {
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	workspace.onDidChangeTextDocument(event => {
		if (activeEditor && event.document === activeEditor.document) {
			triggerUpdateDecorations();
	 	}
	}, null, context.subscriptions);

	workspace.onDidChangeWorkspaceFolders(event => {
		loadLocalizations();
	}, null, context.subscriptions);

	workspace.onDidChangeTextDocument(event => {
		if (event.contentChanges.length <= 0 || !workspace.workspaceFolders || workspace.workspaceFolders.length <= 0) {
			return;
		}

		let primaryFolder = workspace.workspaceFolders[0];
		let localesFolder = joinPath(primaryFolder.uri.fsPath, 'lang');

		let inLocalesFolder = event.document.fileName.startsWith(localesFolder);
		if (!inLocalesFolder) {
			return;
		}
		
		loadLocalizations(() => {
			updateDecorations(true);
		});
	}, null, context.subscriptions);

	workspace.onDidChangeConfiguration(event => {
		loadConfiguration();
	}, null, context.subscriptions);

	loadLocalizations();
	
	context.subscriptions.push(commands.registerCommand('localize.switchLocale', async () => {
		let selectedLocale = await switchLocale(loadedLocales);

		if (!selectedLocale) {
			return;
		}

		currentLocale = selectedLocale;

		loadLocalizations(() => {
			updateDecorations(true);
		});
	}));
}